File: backend/internal/dto/request/experiment.go
package request

import "github.com/chainspace/backend/internal/model"

// CreateExperimentRequest 创建实验请求
type CreateExperimentRequest struct {
	ChapterID     uint                      `json:"chapter_id" binding:"required"`
	Title         string                    `json:"title" binding:"required,max=200"`
	Description   string                    `json:"description" binding:"omitempty"`
	Type          string                    `json:"type" binding:"required,oneof=visualization code_dev command_op data_analysis tool_usage config_debug reverse troubleshoot collaboration"`
	Difficulty    int                       `json:"difficulty" binding:"omitempty,min=1,max=5"`
	EstimatedTime int                       `json:"estimated_time" binding:"omitempty,min=1"`
	MaxScore      int                       `json:"max_score" binding:"omitempty,min=0"`
	PassScore     int                       `json:"pass_score" binding:"omitempty,min=0"`
	AutoGrade     bool                      `json:"auto_grade"`
	Blueprint     model.ExperimentBlueprint `json:"blueprint" binding:"required"`
	SortOrder     int                       `json:"sort_order" binding:"omitempty,min=0"`
	StartTime     string                    `json:"start_time" binding:"omitempty"`
	EndTime       string                    `json:"end_time" binding:"omitempty"`
	AllowLate     bool                      `json:"allow_late"`
	LateDeduction int                       `json:"late_deduction" binding:"omitempty,min=0,max=100"`
}

// UpdateExperimentRequest 更新实验请求
type UpdateExperimentRequest struct {
	ChapterID     *uint                      `json:"chapter_id" binding:"omitempty,min=1"`
	Title         string                     `json:"title" binding:"omitempty,max=200"`
	Description   string                     `json:"description" binding:"omitempty"`
	Type          string                     `json:"type" binding:"omitempty,oneof=visualization code_dev command_op data_analysis tool_usage config_debug reverse troubleshoot collaboration"`
	Difficulty    *int                       `json:"difficulty" binding:"omitempty,min=1,max=5"`
	EstimatedTime *int                       `json:"estimated_time" binding:"omitempty,min=1"`
	MaxScore      *int                       `json:"max_score" binding:"omitempty,min=0"`
	PassScore     *int                       `json:"pass_score" binding:"omitempty,min=0"`
	AutoGrade     *bool                      `json:"auto_grade" binding:"omitempty"`
	Blueprint     *model.ExperimentBlueprint `json:"blueprint" binding:"omitempty"`
	SortOrder     *int                       `json:"sort_order" binding:"omitempty,min=0"`
	StartTime     string                     `json:"start_time" binding:"omitempty"`
	EndTime       string                     `json:"end_time" binding:"omitempty"`
	AllowLate     *bool                      `json:"allow_late" binding:"omitempty"`
	LateDeduction *int                       `json:"late_deduction" binding:"omitempty,min=0,max=100"`
	Status        string                     `json:"status" binding:"omitempty,oneof=draft published archived"`
}

// StartEnvRequest 启动实验环境请求
type StartEnvRequest struct {
	ExperimentID uint   `json:"experiment_id" binding:"required"`
	SnapshotURL  string `json:"snapshot_url" binding:"omitempty,max=500"`
}

// ExtendEnvRequest 延长实验环境时间请求
type ExtendEnvRequest struct {
	Duration int `json:"duration" binding:"required,min=1,max=240"`
}

type CreateSnapshotRequest struct{}

// SubmitExperimentRequest 提交实验请求
type SubmitExperimentRequest struct {
	ExperimentID uint   `json:"experiment_id" binding:"required"`
	EnvID        string `json:"env_id" binding:"omitempty,max=50"`
	Content      string `json:"content" binding:"omitempty"`
	FileURL      string `json:"file_url" binding:"omitempty,max=500"`
}

// GradeSubmissionRequest 批改提交请求
type GradeSubmissionRequest struct {
	Score    int    `json:"score" binding:"required,min=0"`
	Feedback string `json:"feedback" binding:"omitempty"`
}

// ListExperimentsRequest 实验列表请求
type ListExperimentsRequest struct {
	PaginationRequest
	CourseID  uint   `form:"course_id"`
	ChapterID uint   `form:"chapter_id"`
	Type      string `form:"type"`
	Status    string `form:"status"`
	Keyword   string `form:"keyword"`
}

// ListEnvsRequest 实验环境列表请求
type ListEnvsRequest struct {
	PaginationRequest
	ExperimentID uint   `form:"experiment_id"`
	UserID       uint   `form:"user_id"`
	Status       string `form:"status"`
}

type ListExperimentSessionsRequest struct {
	PaginationRequest
	ExperimentID uint   `form:"experiment_id"`
	Status       string `form:"status"`
}

type ListExperimentSessionMessagesRequest struct {
	PaginationRequest
}

type SendExperimentSessionMessageRequest struct {
	Message string `json:"message" binding:"required"`
}

type UpdateExperimentSessionMemberRequest struct {
	RoleKey         *string `json:"role_key" binding:"omitempty,max=50"`
	AssignedNodeKey *string `json:"assigned_node_key" binding:"omitempty,max=50"`
	JoinStatus      *string `json:"join_status" binding:"omitempty,oneof=joined left"`
}

// ListSubmissionsRequest 提交列表请求
type ListSubmissionsRequest struct {
	PaginationRequest
	ExperimentID uint   `form:"experiment_id"`
	StudentID    uint   `form:"student_id"`
	Status       string `form:"status"`
}

// CreateDockerImageRequest 创建镜像请求
type CreateDockerImageRequest struct {
	Name        string                 `json:"name" binding:"required,max=100"`
	Tag         string                 `json:"tag" binding:"required,max=50"`
	Registry    string                 `json:"registry" binding:"omitempty,max=200"`
	Description string                 `json:"description" binding:"omitempty"`
	Category    string                 `json:"category" binding:"omitempty,max=50"`
	Features    []interface{}          `json:"features" binding:"omitempty"`
	EnvVars     map[string]interface{} `json:"env_vars" binding:"omitempty"`
	Ports       []interface{}          `json:"ports" binding:"omitempty"`
	BaseImage   string                 `json:"base_image" binding:"omitempty,max=100"`
}

// UpdateDockerImageRequest 更新镜像请求
type UpdateDockerImageRequest struct {
	Tag         string                 `json:"tag" binding:"omitempty,max=50"`
	Description string                 `json:"description" binding:"omitempty"`
	Category    string                 `json:"category" binding:"omitempty,max=50"`
	Features    []interface{}          `json:"features" binding:"omitempty"`
	EnvVars     map[string]interface{} `json:"env_vars" binding:"omitempty"`
	Ports       []interface{}          `json:"ports" binding:"omitempty"`
	Status      string                 `json:"status" binding:"omitempty,oneof=active inactive"`
}

// ListDockerImagesRequest 镜像列表请求
type ListDockerImagesRequest struct {
	PaginationRequest
	Category  string `form:"category"`
	Status    string `form:"status"`
	IsBuiltIn *bool  `form:"is_built_in"`
	Keyword   string `form:"keyword"`
}

File: frontend/src/types/experiment.ts
import type { ExperimentBlueprint } from './experimentBlueprint'
import type { EnvStatus } from './experimentSession'

export type ExperimentType =
  | 'visualization'
  | 'code_dev'
  | 'command_op'
  | 'data_analysis'
  | 'tool_usage'
  | 'config_debug'
  | 'reverse'
  | 'troubleshoot'
  | 'collaboration'

export type ExperimentStatus = 'draft' | 'published' | 'archived'
export type SubmissionStatus = 'pending' | 'grading' | 'graded'

export interface Experiment {
  id: number
  school_id: number
  course_id?: number
  chapter_id: number
  chapter_title?: string
  creator_id: number
  creator_name?: string
  title: string
  description: string
  type: ExperimentType
  mode: import('./experimentBlueprint').ExperimentMode
  difficulty: number
  estimated_time: number
  max_score: number
  pass_score: number
  auto_grade: boolean
  blueprint: ExperimentBlueprint
  sort_order: number
  status: ExperimentStatus
  start_time?: string
  end_time?: string
  allow_late: boolean
  late_deduction: number
  submission_count?: number
  my_score?: number
  my_status?: string
  created_at: string
}

export interface SubmissionCheckResult {
  id: number
  submission_id: number
  checkpoint_key: string
  checkpoint_type: string
  target?: string
  passed: boolean
  score: number
  details?: string
  sort_order: number
  created_at: string
  updated_at: string
}

export interface Submission {
  id: number
  experiment_id: number
  experiment_title?: string
  student_id: number
  student_name?: string
  env_id?: string
  content?: string
  file_url?: string
  snapshot_url?: string
  score?: number | null
  auto_score?: number | null
  manual_score?: number | null
  feedback?: string
  check_results?: SubmissionCheckResult[]
  status: SubmissionStatus
  submitted_at: string
  graded_at?: string
  grader_name?: string
  is_late: boolean
  attempt_number: number
}

export interface CreateExperimentRequest {
  chapter_id: number
  title: string
  description?: string
  type: ExperimentType
  difficulty?: number
  estimated_time?: number
  max_score?: number
  pass_score?: number
  auto_grade?: boolean
  blueprint: ExperimentBlueprint
  sort_order?: number
  start_time?: string
  end_time?: string
  allow_late?: boolean
  late_deduction?: number
}

export interface UpdateExperimentRequest extends Partial<CreateExperimentRequest> {
  status?: ExperimentStatus
}

export const ExperimentTypeMap: Record<ExperimentType, string> = {
  visualization: '可视化交互类',
  code_dev: '代码开发类',
  command_op: '命令操作类',
  data_analysis: '数据分析类',
  tool_usage: '工具使用类',
  config_debug: '配置调试类',
  reverse: '逆向分析类',
  troubleshoot: '故障排查类',
  collaboration: '多人协作类',
}

export const ExperimentTypeIcon: Record<ExperimentType, string> = {
  visualization: 'eye',
  code_dev: 'code',
  command_op: 'terminal',
  data_analysis: 'bar-chart',
  tool_usage: 'wrench',
  config_debug: 'settings',
  reverse: 'search',
  troubleshoot: 'bug',
  collaboration: 'users',
}

export const EnvStatusMap: Record<EnvStatus, { text: string; color: string }> = {
  pending: { text: '待启动', color: 'default' },
  creating: { text: '创建中', color: 'processing' },
  running: { text: '运行中', color: 'success' },
  paused: { text: '已暂停', color: 'warning' },
  terminated: { text: '已终止', color: 'default' },
  failed: { text: '启动失败', color: 'error' },
}
